kempner_hpc_handbook/scalability/gpu_computing.md

```diff
@@ -1,88 +1,130 @@
 # GPU Computing
 
-## Machine Learning Parallelism Approaches on GPUs
-The scaling of machine learning (ML) workloads in HPC environments can be achieved through various parallelism approaches. This section outlines the primary methods for parallelizing ML computations.
-
-### Data Parallelism
-Involves splitting the dataset into smaller batches that are processed in parallel across different GPUs. Each GPU trains a copy of the model on its subset of the data, and the results are aggregated to update the model.
-
-
-```{figure} figures/png/data_parallel.PNG
----
-height: 450 px
-name: Data Parallelism Diagram
----
-(*Credit: [nvidia.com](https://nvidia.com)*)
-```
-
-### Model Parallelism
-The model's parameters are divided across multiple GPUs. This approach is useful for training large models that cannot fit into the memory of a single GPU. GPUs work on different parts of the model simultaneously.
-
-
-```{figure} figures/png/tensor_parallel_1.PNG
+## GPU vs CPU Computing
+The difference in CPU and GPU computing capabilities comes from their design. CPUs are designed to execute a sequence of instructions (aka thread) as fast as possible and with multicore design to run multiple threads simultaneously, mostly targeting lower response time. While GPU design focuses on providing higher throughput by executing thousands of threads concurrently by providing more computing resources aiming to excel parallel workloads efficiently.
+```{figure} figures/png/gpu_vs_cpu.png
 ---
 height: 300px
-name: Model Parallelism Diagram 1
+name: GPU Compute Resources
 ---
-Model Parallelism Diagram (*Credit: [nvidia.com](https://nvidia.com)*)
+GPU design provides vastly greater compute resources particularly for parallel workloads
 ```
 
-```{figure} figures/png/tensor_parallel_2.PNG
----
-height: 350px
-name: Model Parallelism Diagram 2
----
-Model Parallelism Diagram (*Credit: [nvidia.com](https://nvidia.com)*)
-```
+## How to use GPUs
+Here is how we can utilize the compute power of GPUs for Matrix Multiplication using the CuPy package (equivalent of NumPy for GPUs).
 
-```{figure} figures/png/tensor_parallel_3.PNG
----
-height: 350px
-name: Model Parallelism Diagram 3
----
-Model Parallelism Diagram (*Credit: [nvidia.com](https://nvidia.com)*)
-```
+````{dropdown} Matrix Multiplication Using CuPy on GPU vs NumPy on CPU
+```{code}
+import numpy as np
+import cupy as cp
+import time
```

Comment:
We follow [PEP8](https://peps.python.org/pep-0008/) style guide. So, Python's internal modules should be imported first.